arangod/Aql/Query.cpp

```diff
@@ -1375,15 +1391,62 @@ aql::ExecutionState Query::cleanupTrxAndEngines(ErrorCode errorCode) {
-    _sharedState->executeAndWakeup([&] {
-      _shutdownState.store(ShutdownState::Done, std::memory_order_relaxed);
-      return true;
+  try {
+    TRI_IF_FAILURE("Query::finalize_error_on_finish_db_servers") {
+      THROW_ARANGO_EXCEPTION(TRI_ERROR_INTERNAL_AQL);
+    }
+    ::finishDBServerParts(*this, errorCode).thenValue([ss = _sharedState, this](Result r) {
+      LOG_TOPIC_IF("fd31e", INFO, Logger::QUERIES, r.fail() && r.isNot(TRI_ERROR_HTTP_NOT_FOUND))
+        << "received error from DBServer on query finalization: " << r.errorNumber() << ", '" << r.errorMessage() << "'";
+      _sharedState->executeAndWakeup([&] {
+        _shutdownState.store(ShutdownState::Done, std::memory_order_relaxed);
+        return true;
+      });
     });
-  });
-  
-  return ExecutionState::WAITING;
+    return ExecutionState::WAITING;
+  } catch (...) {
+    // In case of any error that happend in sending out the requests
+    // we simply reset to done, we tried to cleanup the engines.
+    // we only get here if something in the network stack is out of order.
+    // so there is no need to retry on cleaning up the engines, caller can continue
+    // Also note: If an error in cleanup happens the query was completed already,
+    // so this error does not need to be reported to client.
+    _shutdownState.store(ShutdownState::Done, std::memory_order_relaxed);
+
+    if (isModificationQuery()) {
+      // For modification queries these left-over locks will have negative side effects
+      // We will report those to the user.
+      // Lingering Read-locks should not block the system.
+      std::vector<std::string_view> writeLocked{};
+      std::vector<std::string_view> exclusiveLocked{};
+      _collections.visit([&](std::string const& name, Collection& col) -> bool {
+        switch (col.accessType()) {
+          case AccessMode::Type::WRITE: {
+            writeLocked.emplace_back(name);
+            break;
+          }
+          case AccessMode::Type::EXCLUSIVE: {
+            exclusiveLocked.emplace_back(name);
+            break;
+          }
+          default:
+            // We do not need to capture reads.
+            break;
+        }
+        return true;
+      });
+      LOG_TOPIC("63572", WARN, Logger::QUERIES)
+          << " Failed to cleanup leftovers of a query. Due to communication errors. "
+          << " The DBServers will eventually clean up the state. The following locks still exist: "
+          << " write: " << writeLocked << " you may not drop this collection until timeout."
+          << " exclusive: " << exclusiveLocked << " you may not access this collection until timeout.";
+      for (auto const& [serverDst, queryId] : _serverQueryIds) {
+        TRI_ASSERT(serverDst.substr(0, 7) == "server:");
+        auto msg = "Failed to send unlock request DELETE _api/aql/finish/" + std::to_string(queryId) + " to " + serverDst;
```

Comment:
We need to include the database in this message.
Otherwise it will only work on _system

Reply:
Sorted out